### PR TITLE
Implement hash consing for `Sym`

### PR DESCRIPTION
**Pull request by bowenszhu**
- #614

This PR implements hash consing for `Sym` as the first step towards applying hash consing across SymbolicUtils.jl.

### Motivation:

Hash consing enables efficient memory usage and potential performance improvements by ensuring that only one unique copy of a symbolic expression exists in memory.

### Implementation Details:

- A [`WeakValueDict`](https://github.com/twavv/WeakValueDicts.jl) is used to store existing symbolic objects. This choice ensures that symbolic objects without strong references are eligible for garbage collection.
- The `BasicSymbolic` struct is modified to be `mutable` to comply with the requirements of `WeakValueDict` (due to how Julia implements finalizers).
- A new "hash" function, `hash2`, is introduced to incorporate `symtype` into the hash calculation. This approach was chosen over modifying the existing `hash` function to avoid breaking many tests because ordering is dependent on hash values.
- A new `isequal2` function is added for metadata comparison in addition to `Base.isequal`, because directly modifying `Base.isequal` breaks numerous tests in ModelingToolkit.jl.

Future PRs will apply hash consing to other `BasicSymbolic` subtypes.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Looks reasonable enough and the implementation is clean. Are there performance benchmarks to see if it's going on the right direction?

**Comment by AayushSabharwal** [comment]
Considering this matches hashes to symbolic variables, wouldn't it also need to handle hash collisions? If two `Sym`s end up with the same hash, it would return a different one than what the user expects.

Also, if hashconsing is using a custom hash function could it also hash the metadata and not require that it be empty?

**Comment by bowenszhu** [comment]
requesting review @AayushSabharwal @ChrisRackauckas

File: src/types.jl [hunk 1]
```diff
@@ -77,6 +77,8 @@ function exprtype(x::BasicSymbolic)
     end
 end
 
+wvd = WeakValueDict{UInt, BasicSymbolic}()
```
Comment (AayushSabharwal) [inline src/types.jl:80]: Is there a benefit to not making this `const`? Wouldn't it lead to non-const global issues?

Comment (bowenszhu) [reply src/types.jl]: Updated. Thank you!

File: src/types.jl [hunk 2]
```diff
@@ -265,6 +271,26 @@ function _isequal(a, b, E)
...
+and ignores metadata. This does not help deal with hash collisions when metadata is 
+relevant for distinguishing expressions, particularly in hashing contexts. This function
+provides a stricter equality check that includes metadata comparison, preventing
+such collisions.
+
+Modifying `Base.isequal` directly breaks numerous tests in `SymbolicUtils.jl` and
+downstream packages like `ModelingToolkit.jl`, hence the need for this separate
+function.
+"""
+function isequal2(a::BasicSymbolic, b::BasicSymbolic)::Bool
```
Comment (ChrisRackauckas) [inline src/types.jl:290]: needs a better name, `complete_isequal`, `isequal_with_metadata`, or something .

Comment (bowenszhu) [reply src/types.jl]: Changed to `isequal_with_metadata`.

**Pull request merged**